src/DotNetWorker.ApplicationInsights/Initializers/AppServiceEnvironmentVariableMonitor.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Azure.Functions.Worker.ApplicationInsights.Initializers;
+
+internal class AppServiceEnvironmentVariableMonitor : BackgroundService, IOptionsChangeTokenSource<AppServiceOptions>
+{
+    private readonly TimeSpan _refreshInterval;
+
+    private IChangeToken _changeToken;
+    private CancellationTokenSource _cancellationTokenSource = new();
+
+    private readonly Dictionary<string, string?> _monitoredVariableCache = new(StringComparer.OrdinalIgnoreCase);
+
+    public AppServiceEnvironmentVariableMonitor() : this(TimeSpan.FromSeconds(5))
+    {
+    }
+
+    public AppServiceEnvironmentVariableMonitor(TimeSpan refreshInterval)
+    {
+        _refreshInterval = refreshInterval;
+        _changeToken = new CancellationChangeToken(_cancellationTokenSource.Token);
+    }
+
+    public string Name => string.Empty;
+
+    public IChangeToken GetChangeToken() => _changeToken;
+
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            bool changeDetected = false;
+
+            foreach (string envVar in AppServiceOptionsInitializer.EnvironmentVariablesToMonitor)
+            {
+                string? currentVal = Environment.GetEnvironmentVariable(envVar);
+                _monitoredVariableCache.TryGetValue(envVar, out string? cachedVal);
+
+                if (!string.Equals(currentVal, cachedVal, StringComparison.Ordinal))
+                {
+                    changeDetected = true;
+                    _monitoredVariableCache[envVar] = currentVal;
+                }
+            }
+
+            if (changeDetected)
+            {
+                var oldTokenSource = Interlocked.Exchange(ref _cancellationTokenSource, new CancellationTokenSource());
+                Interlocked.Exchange(ref _changeToken, new CancellationChangeToken(_cancellationTokenSource.Token));
+
+                if (!oldTokenSource.IsCancellationRequested)
+                {
+                    oldTokenSource.Cancel();
+                    oldTokenSource.Dispose();
+                }
+            }
+
+            try
+            {
+                await Task.Delay(_refreshInterval, stoppingToken);
```

Comment:
Is it possible that one of these env variables changes more than once (by platform) within the 5 second timespan? In that case, would we miss the second update?

Reply:
I *think* they'll only ever change at most once. I was tempted even to kill the timer once we detected the first change, but I can't confirm that that behavior will never change.

But if they change twice within 5 seconds, we'll see the last change (we'd miss that middle, but that's okay).

This does mean there's a small window where some telemetry would be wrong due to the env variable swap (up to 5 seconds until we notice the change), but that shouldn't be a huge issue. Heck, App Insights defaults to 60 seconds for checking this by default...